java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -236,54 +262,74 @@ public void onCompleted() {
 
         };
 
-    asyncServerStreamingCall(call, ticket.toProtocol(), clientResponseObserver);
+    ClientCalls.asyncServerStreamingCall(call, ticket.toProtocol(), clientResponseObserver);
     return stream;
   }
 
-  private static class SetStreamObserver<T> implements StreamObserver<T> {
-    private final SettableFuture<T> result = SettableFuture.create();
-    private volatile T resultLocal;
+  private static class SetStreamObserver implements StreamObserver<Flight.PutResult> {
+
+    private final CompletableFuture<Void> result;
+    private final StreamListener<PutResult> listener;
+
+    SetStreamObserver(FlightProducer.StreamListener<PutResult> listener) {
+      this.listener = listener;
```

Comment:
I suggest doing a noop listener here if input is null rather than the if(listener != null) down below.